src/plugins/mmapstorage/mmapstorage.c

```diff
@@ -0,0 +1,1323 @@
+/**
+ * @file
+ *
+ * @brief Source for mmapstorage plugin
+ *
+ * @copyright BSD License (see LICENSE.md or https://www.libelektra.org)
+ *
+ */
+#define _XOPEN_SOURCE 600
+
+/* -- Imports --------------------------------------------------------------------------------------------------------------------------- */
+
+#include "mmapstorage.h"
+#include "dynarray.h"
+
+#include <kdberrors.h>
+#include <kdbhelper.h>
+#include <kdblogger.h>
+#include <kdbmmap.h>
+#include <kdbprivate.h>
+
+#ifdef HAVE_KDBCONFIG_H
+#include "kdbconfig.h"
+#endif
+
+#include <assert.h>
+#include <errno.h>
+#include <limits.h>    // SSIZE_MAX
+#include <stdint.h>    // uintN_t, uintptr_t
+#include <stdio.h>     // fopen(), fileno(), fread()
+#include <stdlib.h>    // strtol()
+#include <string.h>    // memcmp()
+#include <sys/mman.h>  // mmap()
+#include <sys/stat.h>  // stat()
+#include <sys/types.h> // ftruncate (), size_t
+#include <unistd.h>    // close(), ftruncate()
+
+#ifdef ENABLE_MMAP_CHECKSUM
+#include <zlib.h> // crc32()
+#endif
+
+/* -- Macros & Definitions -------------------------------------------------------------------------------------------------------------- */
+
+#define SIZEOF_KEY (sizeof (Key))
+#define SIZEOF_KEY_PTR (sizeof (Key *))
+#define SIZEOF_KEYSET (sizeof (KeySet))
+#define SIZEOF_KEYSET_PTR (sizeof (KeySet *))
+#define SIZEOF_MMAPHEADER (sizeof (MmapHeader))
+#define SIZEOF_MMAPMETADATA (sizeof (MmapMetaData))
+#define SIZEOF_MMAPFOOTER (sizeof (MmapFooter))
+#define SIZEOF_ADDR_STRING (19) // format: 0xADDR -> ADDR in hex, for 64bit addr: 2 bytes (0x) + 16 bytes (ADDR) + 1 byte (ending null)
+
+struct _mmapAddr
+{
+	KeySet * const ksPtr;
+	char * metaKsPtr;
+	char * ksArrayPtr;
+	char * metaKsArrayPtr;
+	char * keyPtr;
+	char * dataPtr;
+
+	const uintptr_t mmapAddrInt;
+};
+
+typedef struct _mmapAddr MmapAddr;
+
+static KeySet magicKeySet;
+static Key magicKey;
+
+
+/* -- File handling --------------------------------------------------------------------------------------------------------------------- */
+
+/**
+ * @brief Wrapper for fopen().
+ *
+ * @param parentKey containing the filename
+ * @param mode file open mode
+ *
+ * @return file pointer
+ */
+static FILE * openFile (Key * parentKey, const char * mode)
+{
+	FILE * fp;
+	ELEKTRA_LOG_DEBUG ("opening file %s", keyString (parentKey));
+
+	if ((fp = fopen (keyString (parentKey), mode)) == 0)
+	{
+		ELEKTRA_LOG_WARNING ("error opening file %s", keyString (parentKey));
```

Comment:
warning or error?

Reply:
The plugin returns an error (and sets it in the parentKey), should this happen. The log level is warning because the level error is reserved for `ELEKTRA_ASSERT` according to doc/dev/logging.md.
Maybe I am misunderstanding something here?

Reply:
No, everything is fine! Thank you for clearing this up!